File: recursion/SortStack.java
import java.util.*;

public class SortStack {

    public static void main(String[] args) {


         // Example 1
         Stack<Integer> stack1 = new Stack<>();
         stack1.push(3);
         stack1.push(2);
         stack1.push(1);
         Stack<Integer> sortedStack1 = sort(stack1);
         System.out.println("Sorted Stack 1: " + sortedStack1);  // Output: [3, 2, 1]
 
         // Example 2
         Stack<Integer> stack2 = new Stack<>();
         stack2.push(11);
         stack2.push(2);
         stack2.push(32);
         stack2.push(3);
         stack2.push(41);
         Stack<Integer> sortedStack2 = sort(stack2);
         System.out.println("Sorted Stack 2: " + sortedStack2);  // Output: [41, 32, 11, 3, 2]
    }

    static void  insertSorted(Stack<Integer> stack, int element){
        // base case: if the stack is empty or the top of the stack is less than the element
        if(stack.isEmpty() || stack.peek() < element){
            stack.push(element);
            return;
        }

        // else pop the top element and insert the element recursively
        int top = stack.pop();
        insertSorted(stack, element);

        // insert the popped element back to the stack
        stack.push(top);
    }

    static  void sortStack(Stack<Integer> stack){
        if(stack.isEmpty() || stack.size() == 1){
            return;
        }

        int top = stack.pop();


        sortStack(stack);


        insertSorted(stack, top);
    }

    // Function to sort the stack and return it
    static  Stack<Integer> sort(Stack<Integer> stack) {
        sortStack(stack);
        return stack;
    }
}

File: recursion/sortStack.ts
class Solution {
    // Recursive function to insert an element into the sorted stack
    insertSorted(stack: number[], element: number): void {
        // Base case: if stack is empty or top element is smaller than the current element
        if (stack.length === 0 || stack[stack.length - 1] <= element) {
            stack.push(element);
            return;
        }

        // Otherwise, remove the top element and insert the current element recursively
        const top = stack.pop()!;
        this.insertSorted(stack, element);

        // Put the top element back after insertion
        stack.push(top);
    }

    // Recursive function to sort the stack
    sortStack(stack: number[]): void {
        // Base case: if the stack has one or no elements, it's already sorted
        if (stack.length <= 1) {
            return;
        }

        // Remove the top element
        const top = stack.pop()!;

        // Recursively sort the remaining stack
        this.sortStack(stack);

        // Insert the removed element back in the sorted stack
        this.insertSorted(stack, top);
    }

    // Wrapper function to sort the stack
    sort(stack: number[]): number[] {
        this.sortStack(stack);
        return stack;
    }
}

// Example usage
const solution = new Solution();
const stack1 = [3, 2, 1];
console.log(solution.sort(stack1));  // Output: [3, 2, 1]

const stack2 = [11, 2, 32, 3, 41];
console.log(solution.sort(stack2));  // Output: [41, 32, 11, 3, 2]
